### PR TITLE
SUS-2006: Chat - implement rate limit handling on the client side

### PR DESCRIPTION
**Pull request by mszabo-wikia**
Temporarily disable the input field and show an error placeholder if the user was throttled by the server.

https://wikia-inc.atlassian.net/browse/SUS-2006

### CODE REVIEW

**Comment by WikiaTech** [comment]
SonarQube analysis reported 70 issues
* ![MAJOR][MAJOR] 60 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 8 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Chat_setup.php#L209](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/Chat_setup.php#L209): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L209](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/Chat_setup.php#L209): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L212](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/Chat_setup.php#L212): Replace or remove this deprecated Wikia::log logging statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AWikiaLogUsage)
1. ![MAJOR][MAJOR] [controllers.js#L31](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L31): Remove this logging statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2228)
1. ![MAJOR][MAJOR] [controllers.js#L68](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L68): Remove the unused function parameter "transport". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)
1. ![MAJOR][MAJOR] [controllers.js#L94](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L94): Add a "default" clause to this "switch" statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3ASwitchWithoutDefault)
1. ![MAJOR][MAJOR] [controllers.js#L140](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L140): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L140](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L140): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L144](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L144): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L177](https://github.com/Wikia/app/blob/fdcc432ce608b166015ee6a976475dd0d2560b23/extensions/wikia/Chat2/js/controllers/controllers.js#L177): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: extensions/wikia/Chat2/Chat_setup.php [hunk 1]
```diff
@@ -190,8 +190,12 @@
 		'chat-private-headline',
 		'chat-user-blocked',
 		'chat-user-allow',
+
+	    // misc
+	    'chat-user-throttled',
```
Comment (macbre) [inline extensions/wikia/Chat2/Chat_setup.php]: Mixed tabs and spaces

File: extensions/wikia/Chat2/js/controllers/controllers.js [hunk 2]
```diff
@@ -165,6 +165,7 @@ var NodeRoomController = $.createClass(Observable, {
 		this.socket.bind('longMessage', $.proxy(this.onLongMessage, this));
 
 		this.socket.bind('logout', $.proxy(this.onLogout, this));
+		this.socket.bind('freeze', this.onFreeze.bind(this));
```
Comment (macbre) [inline extensions/wikia/Chat2/js/controllers/controllers.js:183]: Are we sure that `Function.prototype.bind` is present in all browsers that we support? Let's be consistent with the rest of the code here and use jQuery's `bind`.

**Review by macbre — CHANGES_REQUESTED** [review]
See inline comments

**Comment by WikiaTech** [comment]
SonarQube analysis reported 70 issues
* ![MAJOR][MAJOR] 60 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 8 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Chat_setup.php#L209](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/Chat_setup.php#L209): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L209](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/Chat_setup.php#L209): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L212](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/Chat_setup.php#L212): Replace or remove this deprecated Wikia::log logging statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AWikiaLogUsage)
1. ![MAJOR][MAJOR] [controllers.js#L31](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L31): Remove this logging statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2228)
1. ![MAJOR][MAJOR] [controllers.js#L68](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L68): Remove the unused function parameter "transport". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)
1. ![MAJOR][MAJOR] [controllers.js#L94](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L94): Add a "default" clause to this "switch" statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3ASwitchWithoutDefault)
1. ![MAJOR][MAJOR] [controllers.js#L140](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L140): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L140](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L140): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L144](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L144): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L177](https://github.com/Wikia/app/blob/d9a54c5c1bf041c5724ba02923f06d0dd457eb22/extensions/wikia/Chat2/js/controllers/controllers.js#L177): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by ghost — APPROVED** [review]

**Review by macbre — APPROVED** [review]

**Comment by WikiaTech** [comment]
SonarQube analysis reported 69 issues
* ![MAJOR][MAJOR] 59 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 8 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Chat_setup.php#L210](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/Chat_setup.php#L210): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L210](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/Chat_setup.php#L210): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L213](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/Chat_setup.php#L213): Replace or remove this deprecated Wikia::log logging statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AWikiaLogUsage)
1. ![MAJOR][MAJOR] [controllers.js](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js): File "controllers.js" has 1016 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS104)
1. ![MAJOR][MAJOR] [controllers.js#L108](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js#L108): Add a "default" clause to this "switch" statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3ASwitchWithoutDefault)
1. ![MAJOR][MAJOR] [controllers.js#L154](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js#L154): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L154](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js#L154): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L158](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js#L158): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L192](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js#L192): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L263](https://github.com/Wikia/app/blob/52d058d7c814b24004d385f337ef332493dd3512/extensions/wikia/Chat2/js/controllers/controllers.js#L263): Remove the unused function parameter "index". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 69 issues
* ![MAJOR][MAJOR] 59 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 8 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Chat_setup.php#L210](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/Chat_setup.php#L210): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L210](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/Chat_setup.php#L210): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Chat_setup.php#L213](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/Chat_setup.php#L213): Replace or remove this deprecated Wikia::log logging statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AWikiaLogUsage)
1. ![MAJOR][MAJOR] [controllers.js](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js): File "controllers.js" has 1016 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS104)
1. ![MAJOR][MAJOR] [controllers.js#L108](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js#L108): Add a "default" clause to this "switch" statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3ASwitchWithoutDefault)
1. ![MAJOR][MAJOR] [controllers.js#L154](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js#L154): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L154](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js#L154): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L158](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js#L158): Replace "==" with "===". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L192](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js#L192): Replace "!=" with "!==". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AEqEqEq)
1. ![MAJOR][MAJOR] [controllers.js#L263](https://github.com/Wikia/app/blob/fa9d4ba3cc712dd0e06f457f5e79dc19d945c996/extensions/wikia/Chat2/js/controllers/controllers.js#L263): Remove the unused function parameter "index". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by macbre — APPROVED** [review]

**Pull request merged**